config/config/src/index.ts

```diff
@@ -584,12 +585,10 @@ export async function getConfig (
   }
 
   if (pnpmConfig.workspaceDir != null) {
-    if (cliOptions['workspace-packages']) {
-      pnpmConfig.workspacePackagePatterns = cliOptions['workspace-packages'] as string[]
-    } else {
-      const workspaceManifest = await readWorkspaceManifest(pnpmConfig.workspaceDir)
-      pnpmConfig.workspacePackagePatterns = workspaceManifest?.packages
-    }
+    const workspaceManifest = await readWorkspaceManifest(pnpmConfig.workspaceDir)
+
+    pnpmConfig.workspacePackagePatterns = cliOptions['workspace-packages'] as string[] ?? workspaceManifest?.packages
+    pnpmConfig.catalogs = getCatalogsFromWorkspaceManifest(workspaceManifest)
```

Comment:
When I was considering that option, I originally decided against it since this would make the `WorkspaceManifest` interface no longer match the `pnpm-workspace.yaml` file's contents 1:1. I thought it would be better to leave `WorkspaceManifest` as the interface describing the content of the `pnpm-workspace.yaml` before config normalization steps.

One pattern I've seen for this is a `*Raw` prefix. So if we did want to go this route, there would be a separate `WorkspaceManifest` and `WorkspaceManifestRaw` interface. The former would only contains `catalogs` and the later would contain `catalog` plus `catalogs`.

I thought that would get confusing and decided to keep the catalog config normalization in `@pnpm/config`.

Reply:
Will we use the raw object though? In what case? I guess if we'd have a command to update the contents of catalogs. Will we have such command?

Anyway, I think I would still normalize in read workspace manifest. We can change it in the future or add another function for getting the raw object.

Reply:
> Will we use the raw object though? In what case? I guess if we'd have a command to update the contents of catalogs. Will we have such command?

Yeah. I was imagining `pnpm update` should work for cataloged dependencies. Otherwise catalogs would make dependencies harder to manage for some people.

> Anyway, I think I would still normalize in read workspace manifest. We can change it in the future or add another function for getting the raw object.

It feels strange to perform transformation logic inside of `readWorkspaceManifest`. I really like that `readWorkspaceManifest` just reads the workspace manifest (with some validation) and doesn't do anything more today.

I also like that config-specific logic is in `@pnpm/config`.

If we were to create a new `writeWorkspaceManifest` function, I would expect it to write the same object that's read from `readWorkspaceManifest`. If `readWorkspaceManifest` performs data normalization, reading and writing the manifest would cause the manifest to "change" on disk.

```ts
// Would cause changes on disk.
writeWorkspaceManifest(dir, readWorkspaceManifest(dir))
```

Let me know if you see strong reasons to move this to `readWorkspaceManifest` though.